### ISSUE TITLE
Tried to reverse engineer and structure ICP commands from msm9066_capture

### ISSUE THREAD

**Issue by andrew-p0** [mightymos/msm9066_capture#1]
### ICP Handshake sequence
Following sequence is sent to the device immediately after power on to activate ICP mode
| i2c address | mode | ACK/NACK description |
| ------------- | ------------- | ------------- |
| 0x7e | w | NACK |
| 0x7c | w | Keep sending until ACK |
| 0x7d | w | NACK |
| 0x2d | w | NACK |
| 0x7e | r | ACK |

| i2c address | mode | description |
| ------------- | ------------- | ------------- |
| 0x7c  | r | ?? MCU Reset ??
| 0x7e  | w | command register |
| 0x7f  | r/w | data register |

| commands  | parameter(s) | description  |
| --------- | --------- | ---------  |
| 0x02 |  | ?? Read back 0xfc ?? repeated 2 times, see blankcheck_08-20-2023, mcureset_08-20-2023, erasecode_08-20-2023; On my still locked flash it reads back 0xdc,; read_config_while_protected_fails_08-26-2023 returns 0xCC |
| 0x03 |  | erase flash |
| 0x05 | msb lsb   | write flash (len = 0x80; write 2 times) padded with 0xff or 00  |
| 0x06 | msb lsb   | read flash 512 bytes  |
| 0x08 | adr    | write config byte (write 2 times)  |
| 0x09 | adr    | read config byte  |
| ?? |  | blank check |
| ?? |  | verify |
| ?? |  | protect |
| ?? |  | mcu reset |
| ?? |  | eeprom read |
| ?? |  | eeprom write |
| ?? |  | eeprom erase |

| config byte(s) | parameter(s) | description  |
| --------- | --------- | ---------  |
| address | data |  |
| 0x00  |  0x0a | read chip type OB38S003 = 0x0a|
| 0x10 | 0x00 | chip protect |
| 0x11  |  0xff | system clock  |
|  | 0xf4 |  external crystal (xtal)  |
|  | 0xf5 | ext xtal / 2  |
|  | 0xf6 | ext xtal / 4  |
| | 0xe7 | ext xtal=i/o  |
| | 0xff | 16 MHz  |
| | 0xf9 | 8 MHz  |
| | 0xfa | 4 MHz  |
| | 0xfb | 2 MHz  |
| | 0xfc | 1 MHz  |
                
| 0x12 | 0xf9 | pin func RESET  |
| --------- | --------- | ---------  |
|  | 1111 1xxx | RESET  |
|  | 0000 0xxx | bi-direction IO  |
|  |  | **IO Mode** |
|  | xxxx xxx1 | quasi-bidirection  |
|  | xxxx xxx0 | input only  |
|  |  | **deglitch: XTAL, IRC** |
|  | xxxx x00x | no  |
|  | xxxx x01x | 14ns, 8ns  |
|  | xxxx x10x | 17ns, 0ns  |
|  | xxxx x11x | 20ns, 8ns  |

| config byte(s) | parameter(s) | description  |
| --------- | --------- | ---------  |
| 0x13 | 0xff | reset time  |
|  | 0x06 | 200 ms  |
|  | 0x05 | 100 ms  |
|  | 0x04 | 50 ms  |
|  | 0xff | 25 ms  |
|  | 0x03 | 16 ms  |
|  | 0x02 | 8 ms  |
|  | 0x01 | 4 ms  |
                
| 0x14 | 0xff | ISP Start Addr  |
| --------- | --------- | ---------  |
|  | 0xff | No ISP Code  |
|  | 0x01 | 1f80  |
|  | 0x02 | 1f00  |
|  | 0x03 | 1e80  |
|  | 0x04 | 1e00  |
|  | 0x05 | 1d80  |
|  | 0x06 | 1d00  |
|  | 0x07 | 1c80  |
|  | 0x08 | 1c00  |

| 0x15 | 0xff | ISP Entry Mech.  |
| --------- | --------- | ---------  |
|  | 0xff | default  |
|  | 1xxx xxxx | Addr.($0000)=FFH Power On Disable = 1 / Enable = 0  |
|  | x1xx xxxx | Addr.($0000)=FFH Pad Reset Disable = 1 / Enable = 0  |
|  | xxxx 1xxx | P1.6=0  Power On Disable = 1 / Enable = 0  |
|  | xxxx x1xx | P1.6=0  Pad Reset Disable = 1 / Enable = 0  |
|  | xxxx xx1x | P1.1 inputs 2 clocks    Power On Disable = 1 / Enable = 0  |
|  | xxxx xxx1 | P1.1 inputs 2 clocks    Pad Reset Disable = 1 / Enable = 0  |
|  | xx1x xxxx | Force Entry     Disable = 1 / All resets = 0  |

| 0x16 | 0xff | Low Voltage Reset  |
| --------- | --------- | ---------  |
|  | 0xfc | 3.0V  |
|  | 0xfd | 2.4V  |
|  | 0xfe | 2.2V  |
|  | 0xff | 2.0V  |

| 0x17 | 0xff | Op/Comparator  |
| --------- | --------- | ---------  |
|  | 1111 xxxx | Type 0: GPIO  |
|  | 1010 xxxx | Type 0: Comparator Input 0: Ext pin  |
|  | 1001 xxxx | Type 0: Comparator Input 0: Int VBG |
|  | xxxx 1111 | Type 1: GPIO  |
|  | xxxx 1010 | Type 1: Comparator Input 1: Ext pin  |
|  | xxxx 1001 | Type 1: Comparator Input 0: Int VBG |

| --------- | --------- | ---------  |
| 0x28 | 0xff | ?  |
| 0x29 | 0xff | ?  |
| 0x2a | 0xff | ?  |
| 0x2b | 0xff | ?  |

| --------- | --------- | --------- |
| 0x2c | 00x0 | Flash Checksum 3  |
| 0x2d | 0x1f | Flash Checksum 2  |
| 0x2e | 0x85 | Flash Checksum 1  |
| 0x2f | 0xdb | Flash Checksum 0  |
Checksum is calculated as a SUM of all FLASH bytes. Same is for EEPROM.

| --------- | --------- |
| 0x30 | 0xff | ?  |
| ... | 0xff | ?  |
| 0x3f | 0xff | ?  |

**Comment by mightymos** [comment]
Thank you so much for starting this!
I converted the entries into a github table format so it is easier to read:  
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-tables

I am not sure if these are correct, but I will try to help confirm if I have time.
As best I know with I2C the NACK/ACK is only ever a single bit, so a byte must represent something else.

**Comment by andrew-p0** [comment]
It seems SMAP flasher try to probe address 0x7e (wr) twice in the beginning and getting NACK, meaning device not present. Then there is a delay ~18ms and a series of probes of 0x7f (wr) (ACK), 0x7c (wr) (ACK), 0x7d (wr) (NACK), 0x2d (wr) (NACK), 0x7e (rd) (ACK). I'm not sure if probing 0x7d and 0x2d serves any purpose, like unlocking the interface or something else.

I do not have MSM9066 flasher, so I have tried to connect Sonoff board to my test HW acting as i2c master, but have not succeed in enabling ICP.

**Comment by mightymos** [comment]
@andrew-p0 I have attached individual captures of erase code, blank check, and mcu reset.

I think the NACKs from the other capture might possibly have been due to uart to usb being also connected to microcontroller and interfering with flashing at the start.

If you could at least get erase code to work, and then move on to do blank check, then it would be known that something is working with your HW.

Out of curiosity, what hardware are you using?

**Comment by andrew-p0** [comment]
I've checked the new captures. They all have the same prologue that does not exist in erase_blankcheck_program_verify_mcureset_blink_ihx_02-23-2023a.
All 3 starts with
| cmd (wr) 0x7e | data (rd) 0x7f | cmd (wr) 0x7e | data (rd) 0x7f |
| ------------- | ------------- | ------------- | ------------- |
| 0x02 | 0xfc | 0x02 | 0xfc |

then after ~1ms it seems the status is checked with 
| cmd (wr) 0x7e | data (rd) 0x7f |
| ------------- | ------------- |
| 0x09 0x00 | 0x0a |

The erase command is 0x03 with the clock stretching. The MCU holds SCK low until command is complete.
There is no verify command. Verification is done by reading flash content back with (0x06 msl lsb) command
Blank check is the same. The whole flash is read back and verified for 0xff I assume.
Looks like MCU reset is done by reading i2c address 0x7c.
There are also must be commands for EEPROM erase/write/read.

I was trying to use Raspberry Pi like dev board with i2cdetect, i2cget, i2cset Linux tools, but didn't succeed. I can't get ACK from any addresses that MCU uses. Thinking to try FT2232H mini with Python. Should be easier to control i2c master and add delays if needed. 
Could you try to capture handshaking between MSM9066 and MCU starting from power off state to erase or MCU reset? It could help to identify if there is any handshaking procedure.

Thanks.

**Comment by mightymos** [comment]
I started with Power Off in SMAP, then start capture and click Power On followed by Erase Code.
Uploaded new traces.

I noticed in the SMAP output log that there is a report that chip type is correct.
So there must be some read to determine chip type also.

**Comment by mightymos** [comment]
I uploaded a new trace which also monitors the 3.3V power supply rail.
erasecode_with_3.3V_on_D2_08-22-2023

What makes this more complicated is that I am concerned that the handshake procedure might require controlling power to the target in order to reset the processor (or sending reset command following dummy writes?).

**Comment by mightymos** [comment]
The MSM9066_UserManual_ENG.pdf says:
```
11.1 Please connect OCI_SCL, OCI_SDA, VCC, GND with MSM9066. Press “Autorun” then SMAP
will start ICP programming.
11.2 If OCI_SCL、OCI_SDA as GPIO, that caused enter ICP mode failed. As long as enter ICP
mode during power-on.
(1) Please power off your target board.
(2) Press “Autorun” then SMAP will display “Detecting… …(0%)”. Please reset MCU or power on
your target board, SMAP will start programming.
(3) When finish programming, SMAP shows “MCU Reset Succeeded!”.
```

However, when I power sonoff receiver from 3.3V bench supply (instead of using 3.3V connection from flasher; supply turned off until after clicking Autorun in flasher software), the SMAP flasher still does not detect target.

**Comment by mightymos** [comment]
Ok, as long as the vcc connector of the MSM9066 flasher is connected to target for reading presence of supply voltage (but supply power is set to NO in SMAP!), a bench supply can be used to power and flash microcontroller. Bench supply must be turned on after pressing 'Autorun' in SMAP flasher on computer.

Here is output from SMAP:
```
Initialize finished!
Detecting...
Detect finished!
The chip type is correct.
Erasing...
Erase finished!
The chip type is correct.
```

The new uploaded trace is named:  
erasecode_bench_supplies3.3V_turnedon_08-22-2023.sr

**Comment by andrew-p0** [comment]
> I uploaded a new trace which also monitors the 3.3V power supply rail. erasecode_with_3.3V_on_D2_08-22-2023
> 
> What makes this more complicated is that I am concerned that the handshake procedure might require controlling power to the target in order to reset the processor (or sending reset command following dummy writes?).

[1]. It seems if the "Supply Power" in SMAP is not enabled, than flasher is monitoring VCC line and sending handshake sequence only after VCC goes high.
[2] If flasher is controlling VCC supply, then it tries to send a command to 0x7e address first. After receiving two NACKs it performs power cycle and send a handshake sequence as in [1]. Then retry the command.

**Comment by mightymos** [comment]
I have been able to read chip type and portions of flash memory from a spare Sonoff board.
I used an ESP32 programmed in the Arduino environment.

It needs a lot of work to finally be useful, but I am able to read data that matches values seen in traces.
[esp32_i2c_onbright_flasher.ino](https://github.com/mightymos/msm9066_capture/blob/main/esp32_i2c_onbright_flasher.ino)

Here's example output:
```
high-to-low transition
reset state to idle
low-to-high transition
NACK on transmit of address
NACK on transmit of address
NACK on transmit of address
NACK on transmit of address
Success
ACK detected
error (or nack) count: 3
NACK on transmit of address
NACK on transmit of address
chip type: A
Erasing chip...
chip type: A
FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,FF,
```

**Comment by andrew-p0** [comment]
I have made some progress too. I've been able to send handshake, followed by 0x02 command and read back response. My sonoff board still has an original FW and flash is locked to read:

VCC Detected
Send Handshake
Setup Write to [0x7E} + NAK
Setup Write to [0x7C} + ACK
Setup Write to [0x7D} + NAK
Setup Write to [0x2D} + NAK
Setup Read to [0x7E} + ACK
Send Command
Setup Write to [0x7E} + ACK
0x02 + ACK
Setup Read to [0x7F} + ACK
0xDC + NAK

I have been using Python pyftdi library and FT2232H mini module. Wiring is similar to https://eblot.github.io/pyftdi/api/i2c.html#wiring with the addition of VCC detection on ADBUS3 pin pulled down with 10K resistor.

**Comment by mightymos** [comment]
Great. It seems nice because Python can be used as a flasher on the PC side directly.
It would be nice to have a dump of the original firmware...but I have no idea how to attempt defeating protection.

EDIT: Attached traces for Erase followed by Protect.

**Comment by andrew-p0** [comment]
> EDIT: Attached traces for Erase followed by Protect.
Ok, so protection is done by setting config byte 0x10 to 0. Could you try dump all config bytes on unprotected chip? On protected chip I cant send 0x09 command (NAK). Would be interesting to try to write 0x10 config byte to unprotect chip and keep content. Though I doubt that it can be that simple.

**Comment by mightymos** [comment]
Ok, attached trace of read config while unprotected and attempt to read while protected.

**Comment by andrew-p0** [comment]
Not with the SMAP. What I meant is to use ESP32 since it probably wont work with SMAP flasher. Could you read config byte 0x10 ( command 0x09 0x10) on unprotected chip, and than try to disable protection on protected chip by reverting config byte 0x10 (command 0x08 0x10) to the value that you read from unprotected chip. 
Config byte 0x10 seems to hold flash read protection flag.

**Comment by mightymos** [comment]
I need to do a better test, but it appears that I am able to read config, set protect by writing config, then overwrite config.
However, I think it might not really be working.
Firstly, I think I need to write twice, yet appears to work with a single write in my output?
Secondly, I think the protect/lock might only be enforced after power cycling?

EDIT: If I remove power after locking, I cannot read chip type again until I issue an erase command.

```
Read chip type: 
chip type: A
Reading config...
55
Set protect...
Reading config...
0
Write config...
Reading config...
55
Read chip type: 
chip type: A
```

**Comment by andrew-p0** [comment]
I'm not sure why all write operations are performed 2 times. 
I think that the power cycling is the key. If you'll be able to reset protection after power cycling that would be great.

**Comment by andrew-p0** [comment]
> EDIT: If I remove power after locking, I cannot read chip type again until I issue an erase command.
 So you 
1. lock the chip, 
2. power cycle, 
3. write 0x55 to config 0x10, 
4. power cycle and still can't read chip id?

EDIT: What it returns on command 0x02 when you can't read chip id? Looks like command 0x02 returns some sort of state of the protection and something else. On my board with protection and the original FW it return 0xdc

**Comment by mightymos** [comment]
Most commands do not complete after locking chip, for example cannot read chip id after power cycle.
Command 0x02 does still work with chip protected.
I guess maybe commands less than three might still work.

Command 0x02 returns 0xec or sometimes 0xfc immediately after erasing chip.
Command 0x02 returns 0xcc when chip is protected.

**Comment by mightymos** [comment]
I'm considering ordering an FT module, which one do you have?  
https://www.aliexpress.com/w/wholesale-FT2232H-mini-module.html
https://www.mouser.com/ProductDetail/FTDI/FT2232H-MINI-MODULE?qs=pB3G9VbQXIf%252BpWyngo5ZjA%3D%3D

There is discussion of bypasses readout mechanisms here, but it can be simple or complicated it seems:  
https://research.nccgroup.com/wp-content/uploads/2020/02/NCC-Group-Whitepaper-Microcontroller-Readback-Protection-1.pdf

**Comment by mightymos** [comment]
There is mention in the On Bright datasheet of having an ICE and OCD, but there is not much discussion of them.
It seems like if there is an on board debugger, we should be able to control the CPU externally.
Possibly while holding the CPU in RESET with exposed pad.

https://github.com/arendst/Tasmota/files/9624969/ISSFD-M123_A_OB38S003-Datasheet.EN.-20171106.pdf

**Comment by mightymos** [comment]
Successfully flashed a simple blink program using this sketch.
However, the sketch is rather limited at the moment.
Additionally work could be applied at this project:

https://github.com/mightymos/OnbrightFlasher

**Comment by mightymos** [comment]
@andrew-p0 I cannot figure out how the configuration/fuse byte checksum is being calculated. Were you able to actually calculate it yourself and match to SMAP AC reported checksum?

**Comment by andrew-p0** [comment]
> @andrew-p0 I cannot figure out how the configuration/fuse byte checksum is being calculated. Were you able to actually calculate it yourself and match to SMAP AC reported checksum?

Are you talking about flash checksum that is stored in config bytes 0x2c - 0x2f ? If I remember correctly it was just as SUM of all bytes in flash memory, including blank 0xff bytes.

**Comment by mightymos** [comment]
> > @andrew-p0 I cannot figure out how the configuration/fuse byte checksum is being calculated. Were you able to actually calculate it yourself and match to SMAP AC reported checksum?
> 
> Are you talking about flash checksum that is stored in config bytes 0x2c - 0x2f ? If I remember correctly it was just as SUM of all bytes in flash memory, including blank 0xff bytes.

No sorry, the SMAP AC program on the computer also gives a checksum for the configuration bytes themselves. However, I am having a tough time figuring out how it computes that particular checksum, it does not seem to be simply based on the sum. I have an initial idea that it might be XOR of bytes but am not sure.

**Comment by andrew-p0** [comment]
Oh, OK. Sorry, I have not looked into that since the reverse engineering of
i2c flashing protocol that we did in August of last year. I have been busy
with other household projects since then.
